server/src/lib/jobs/resolve.cpp

```diff
@@ -150,95 +192,173 @@ static std::string byLocation(tnt::Connection& conn, const Group::Condition& con
-            a.keytag='hostname.1' AND e.id_type = {type} AND
-            a.value {op} '{val}')";
-    if (cond.op == Group::ConditionOp::IsNot) {
-        sql = R"(
-                SELECT
-                    id_asset_element
-                FROM
-                    t_bios_asset_element
-                WHERE
-                     id_type = {type} AND id_asset_element NOT IN ()" +
-              sql + ")";
-    }
-    return fmt::format(sql, "type"_a = persist::DEVICE, "op"_a = cond.op != Group::ConditionOp::IsNot ? op(cond) : "=",
-        "val"_a = value(cond));
+    auto byDevice = [&]() {
+        std::string sql = R"(
+            SELECT e.id_asset_element
+            FROM t_bios_asset_element AS e
+            LEFT JOIN t_bios_asset_ext_attributes a ON e.id_asset_element = a.id_asset_element
+            WHERE a.keytag='hostname.1' AND e.id_type = {type} AND
+                  a.value {op} '{val}')";
+
+        if (cond.op == Group::ConditionOp::IsNot) {
+            sql =
+                "SELECT id_asset_element FROM t_bios_asset_element \
+                   WHERE id_type = {type} AND id_asset_element NOT IN (" +
+                sql + ")";
+        }
+
+        // clang-format off
+        return fmt::format(sql,
+            "type"_a = persist::DEVICE,
+            "op"_a   = cond.op != Group::ConditionOp::IsNot ? op(cond) : "=",
+            "val"_a  = value(cond)
+        );
+        // clang-format on
+    };
+
+    auto byVirtDevice = [&]() {
+        std::string sql = R"(
+            SELECT l.id_asset_device_dest FROM t_bios_asset_link AS l
+            LEFT JOIN t_bios_asset_ext_attributes AS a ON a.id_asset_element = l.id_asset_device_src
+            LEFT JOIN t_bios_asset_element AS e ON e.id_asset_element = l.id_asset_device_dest
+            WHERE
+                l.id_asset_link_type IN ({linkTypes}) AND
+                a.keytag = 'hostName' AND
+                a.value {op} '{val}' AND
+                e.id_type = {type} AND
+                e.id_subtype = {subtype}
+        )";
+
+        if (cond.op == Group::ConditionOp::IsNot) {
+            sql =
+                "SELECT id_asset_element FROM t_bios_asset_element \
+                  WHERE id_type = {type} AND id_subtype = {subtype} AND id_asset_element NOT IN (" +
+                sql + ")";
+        }
+
+        // clang-format off
+        return fmt::format(sql,
+            "linkTypes"_a = fty::implode(vmLinkTypes(), ", "),
+            "val"_a       = value(cond),
+            "op"_a        = cond.op != Group::ConditionOp::IsNot ? op(cond) : "=",
+            "type"_a      = persist::VIRTUAL_MACHINE,
+            "subtype"_a   = persist::VMWARE_VM
+        );
+        // clang-format on
+    };
+
+    return "{} union {}"_format(byDevice(), byVirtDevice());
 }
 
+// =====================================================================================================================
+
 static std::string byIpAddress(const Group::Condition& cond)
```

Comment:
If I get it correctly, you are supporting IP list like "192.168 | 10.120.12" ?

Reply:
yes
And by mask